Create ProtobufferBuilder

nityanandaz opened the pull request:
# Create ProtobufferBuilder

## :bulb: Proposed solution

ProtobufferBuilder is organized in a separate target.
Its functionality is developed independently of Apodini.

### Problem that is solved

`ProtobufferSemanticModelBuilder: SemanticModelBuilder, InterfaceExporter` exposes a `.proto` definition of the `Apodini.WebService` at `/apodini/proto/`

### Implications

Only a tiny API is exposed: `ProtobufferBuilder.addService`

## :heavy_plus_sign: Additional Information

### Testing

Are there tests included? Yes, of course.

Which situations are tested and which corner cases are missing?
Different Message kinds are tested and even `Apodini.WebService`s.

Enums are currently not supported.
Thus, there might arise new corner cases in the future.

### Reviewer Nudging

A reviewer might only look at the changes in `Sources/Apodini` first.

nityanandaz committed:
Create targets  GRPCProtobufBuilder(Tests)

nityanandaz committed:
Spell out generic type names

I.e., Array<Node<Int>> becomes ArrayOfNodeOfInt

nityanandaz committed:
Supplement

nityanandaz committed:
Implement _contextMap_

nityanandaz committed:
Rename tests and add test

nityanandaz committed:
Create ProtobufferBuilder

nityanandaz committed:
Supplement

nityanandaz committed:
Implement Tree.map ...again

nityanandaz committed:
Filter appropriate Tree<TypeInfo>.children

nityanandaz committed:
Remove GRPC prefix

nityanandaz committed:
Refactor spellOutGeneric -> compatibleGenericName

nityanandaz committed:
Rename target GRPCProtobufBuilder to ProtobufferBuilder

nityanandaz committed:
Add Property.isRepeated

nityanandaz committed:
Create Tree.edit

Now, testTypeTwoLevelsDeep reaches Transaction.

moritzschuell commented on Tests/ProtobufferBuilderTests/ProtobufferBuilderTests.swift:
```diff
@@ -0,0 +1,86 @@
+}
+
+extension ProtobufferBuilderTests {
+    func testTypeTwoLevelsDeep() throws {
```
Not quite sure I understand the output of this case 🤔
Why does this output the following?
```protobuf
message ArrayOfTransaction {
	_ArrayBufferOfTransaction _buffer = 0;
}
message Transaction {
	Int amount = 0;
}
message Account {
	repeated ArrayOfTransaction transactions = 0;
}
```
I would've expected something like this:
```protobuf
message Transaction {
	Int amount = 0;
}
message Account {
	repeated Transaction transactions = 0;
}
```

moritzschuell commented on Tests/ProtobufferBuilderTests/ProtobufferBuilderTests.swift:
```diff
@@ -0,0 +1,86 @@
+        XCTAssertNoThrow(try code(Account.self))
+    }
+    
+    func testRecursiveType() throws {
```
Similar to the comments above.
Wouldn't the intuitive output be
```protobuf
message NodeOfInt {
	Int value = 0;
	repeated NodeOfInt children = 1;
}
```

moritzschuell commented on Sources/ProtobufferBuilder/Model.swift:
```diff
@@ -0,0 +1,41 @@
+// MARK: - Service and Service.Method
+
+struct Service: Equatable, Hashable {
+    struct Method: Equatable, Hashable {
```
`input` and `output` could each also be a `stream` instead of a single message.
I guess this goes too far for this first PR, but we will also need to support client and server side stream components (as defined in [Apodini's ComponentTypes docu](https://github.com/Apodini/Apodini/tree/feature/parameter/Documentation/ComponentTypes)).

nityanandaz replied:
I agree that this is out of scope. I created a new issue https://github.com/Apodini/Apodini/issues/44

moritzschuell commented on Sources/ProtobufferBuilder/ProtobufferBuilder.swift:
```diff
@@ -0,0 +1,54 @@
+}
+
+public extension ProtobufferBuilder {
+    func add<T>(_ messageType: T.Type = T.self) throws {
```
Currently, this ignores any functions the given type might have.
We will also need to check whether the given type conforms to `Component`, and if so populate the `private var services` (that already exists in `ProtobufferBuilder`) with the component and its `handle` method. 

This closely relates to my comment above, as here we would also have to distinguish between the different types of messages (`stream`s or unary) that the method takes as input and output. 

So, throwing in this 
```swift
struct Person: Component {
    let name: String
    let age: Int32

    public func handle() -> String {
        return "Hello \(name)."
    }
}
```
should produce a service like
```protobuf
service PersonService {
    rpc handle (Person) returns (/*...*/);
}

message Person {
    string name = 1;
    int32 age = 2;
}
```

nityanandaz replied:
Conformance to `Component` is inforced in `ProtobufferSemanticModelBuilder`.
`ProtobufferBuilder.services` is now populated.

I created a follow-up issue: https://github.com/Apodini/Apodini/issues/51

nityanandaz committed:
Create KeyPathSorted

nityanandaz committed:
Extend ParticularType

nityanandaz committed:
Manage expectations

https://github.com/Apodini/Apodini/pull/33#discussion_r534744532

nityanandaz committed:
Actually fix array

nityanandaz committed:
Format output

nityanandaz committed:
Group extensions

nityanandaz committed:
Refactor TypeInfo.compatibleName

nityanandaz committed:
Merge branch 'develop' into create-GRPCProtobufBuilder

nityanandaz committed:
Merge branch 'develop' into create-GRPCProtobufBuilder

nityanandaz committed:
More specific test

https://github.com/Apodini/Apodini/pull/33#discussion_r534747244

nityanandaz committed:
Code style

nityanandaz committed:
Change supportedPrimitiveTypes

Create testScalarType
Create ProtobufferBuilder.addService

nityanandaz committed:
Create ProtobufferSemanticModelBuilder

And create testPokemonWebService

nityanandaz committed:
More specific test

nityanandaz committed:
Create ProtobufferBuilderIntegrationTests

nityanandaz committed:
Swift lint

nityanandaz committed:
Update testTypeTwoLevelsDeep

nityanandaz committed:
Add Message postfix

nityanandaz committed:
Remove level of indirection

nityanandaz committed:
Merge branch 'develop' into create-GRPCProtobufBuilder

nityanandaz committed:
Supplement

nityanandaz committed:
Manage expectations

Create new issue https://github.com/Apodini/Apodini/issues/48

nityanandaz committed:
Lowercase scalar types

nityanandaz committed:
Import FoundationNetworking

Linux (focal) build error: 'URLSession' is unavailable: This type has moved to the FoundationNetworking module. Import that module to use it.

nityanandaz committed:
SwiftLint

nityanandaz committed:
Merge branch 'develop' into create-GRPCProtobufBuilder

nityanandaz committed:
Conditionally import FoundationNetworking

nityanandaz committed:
Update test

nityanandaz committed:
Fulfill test

nityanandaz committed:
SwiftLint

nityanandaz committed:
Add VoidMessage to testPokemonWebService

nityanandaz committed:
Supplement

nityanandaz committed:
Merge branch 'develop' into create-GRPCProtobufBuilder

nityanandaz committed:
Skip test for Linux

nityanandaz committed:
Try higher timeout threshold for IntegrationTest

nityanandaz committed:
Merge branch 'develop' into create-GRPCProtobufBuilder

moritzschuell commented on Sources/ProtobufferBuilder/ProtobufferBuilder.swift:
```diff
@@ -0,0 +1,105 @@
+            return
+        }
+        
+        let serviceName = try serviceNode.value.compatibleName() + "Service"
```
As far as I see, currently only one service per `Component` type can be added, right?
So, if we have two `Text`s in the `content` of the service, only one `TextService` will show up in the `.proto` IDL.
I think this is fine for now, but we should think about generating the service names from the `PathComponent`s (or the `.absolutePath` of an `EndpointsTreeNode` in the case of the shared semantic model).

nityanandaz replied:
Services are now initialized with the absolute path

moritzschuell commented on Sources/ProtobufferBuilder/Extensions/Model+TypeInfo.swift:
```diff
@@ -0,0 +1,48 @@
+                            isRepeated: ParticularType(typeInfo.type).isArray,
+                            name: element.name,
+                            typeName: typeName,
+                            uniqueNumber: offset
```
The first field number / unique number will be 0.
Protobuffer field numbers have to be positive integers.

moritzschuell reviewed (commented):
Looks great so far! 🎉

I noticed that properties with the `@Parameter` property-wrapper applied cause the builder to not generate a service for the component. E.g.
```swift
struct Greeter: Component {
    @Parameter
    var name: String
        
    func handle() -> String {
        return "Hello \(name)"
    }
}
    
var content: some Component {
    Greeter()
}
```
produces only
```proto
syntax = "proto3";
```
While removing the `@Parameter` produces correct output.

nityanandaz committed:
Property.uniqueNumber: increment offset by 1

nityanandaz committed:
Add testGreeterWebService

Ref.: https://github.com/Apodini/Apodini/pull/33#pullrequestreview-549538988

nityanandaz committed:
Change IntegrationTest<GreeterWebService>.exepectedResponse

ProtobufferBuilder.Exception(message: "Tuple: (UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8) is not supported")

Tuple type can be traced back to: public typealias __darwin_uuid_t = (UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8)

nityanandaz committed:
Move fixArray

lschlesinger commented on Sources/ProtobufferBuilder/Extensions/Model+TypeInfo.swift:
```diff
@@ -0,0 +1,48 @@
+                        )
+                    } catch {
+                        print(error)
+                        return nil
```
not sure if "ignoring" an erroneous property is the way to go? Maybe throw an exception or check with precondition?

nityanandaz replied:
I haven't actually encountered erroneous properties, but I have encountered two types of Errors here:

- `couldNotGetTypeInfo(type: Builtin.Int8, kind: Runtime.Kind.opaque)`
- `couldNotGetTypeInfo(type: Builtin.BridgeObject, kind: Runtime.Kind.opaque)`

I use this mechanism to find an _organic_ end to my tree depth.
The above errors are thrown, if Runtime inspects internal implementations of `Swift.String`, for example.

Sadly, I cannot specialize the catch statement with pattern matching, because Runtime.RuntimeError is internal.

lschlesinger commented on Sources/ProtobufferBuilder/Extensions/ProtobufferBuilder+CustomStringConvertible.swift:
```diff
@@ -0,0 +1,30 @@
+        let protoFile = [
+            #"syntax = "proto3";"#,
+            services,
+            messages
```
I am not that familiar with .proto files but I would expect messages before services, as these types and used in the signatures.

nityanandaz replied:
The order doesn't matter. I put services first, because I found them more relevant.

lschlesinger commented on Sources/ProtobufferBuilder/ProtobufferBuilder.swift:
```diff
@@ -0,0 +1,105 @@
+        let voidMessage = Message.void
+        let method = Service.Method(
+            name: "handle",
+            input: voidMessage,
```
just for my understanding, this is where the list of parameters will be used in the future and voidMessage is just temporary?

nityanandaz replied:
Yes, now 0 or 1 Parameters are handled.

I created a follow-up issue: https://github.com/Apodini/Apodini/issues/62

lschlesinger commented on Tests/ProtobufferBuilderTests/ProtobufferBuilderIntegrationTests.swift:
```diff
@@ -0,0 +1,49 @@
+                }
+
+                message PokemonMessage {
+                  int64 id = 0;
```
Not sure if "internal" properties (that are not annotated with `@Parameter` and therefore irrelevant for requests) should be part of the proto message types?

nityanandaz replied:
I filter Parameters now 👍

lschlesinger commented on Sources/Apodini/Semantic Model Builder/gRPC/ProtobufferSemanticModelBuilder.swift:
```diff
@@ -0,0 +1,26 @@
+    
+    override func register<C>(component: C, withContext context: Context) where C : Component {
+        print(Result(catching: { try builder.addService(of: C.self) }))
+        print(Result(catching: { try builder.addMessage(of: C.Response.self) }))
```
I guess you will replace the return type of `handle()` by the type ultimately returned - after being modified through transformers?

nityanandaz replied:
ProtobufferSemanticModelBuilder is now an InterfaceExporter and uses `Endpoint.responseType`

lschlesinger commented on Sources/ProtobufferBuilder/Model.swift:
```diff
@@ -0,0 +1,35 @@
+
+struct Message: Equatable, Hashable {
+    struct Property: Equatable, Hashable {
+        let isRepeated: Bool
```
are you planning to add `optional` in the future?

nityanandaz replied:
I do now 🙂

lschlesinger reviewed (commented):
Looks generally good to me, tried to provide useful feedback :)

nityanandaz committed:
Create EnrichedInfo

nityanandaz committed:
Implement Tree.contextMap

nityanandaz committed:
Increment EnrichedInfo.propertiesOffset

nityanandaz committed:
Create testTypeOneLevelDeep

nityanandaz committed:
Extend Model with EnrichedInfo

nityanandaz committed:
Supplement

nityanandaz committed:
Fix Tree.filter

nityanandaz committed:
Update EnrichedInfo

nityanandaz committed:
Update testRecursiveType

nityanandaz committed:
Create fixPrimitiveTypes

nityanandaz committed:
Delete Message.init(typeInfo: TypeInfo)

Fix fixArray...

nityanandaz committed:
Merge branch 'refactor-Model+TypeInfo' into protobuffer-generation/inspect-component-properties

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/inspect-component-properties

nityanandaz committed:
Implement Tree.compactMap

nityanandaz committed:
Delete IntegrationTest

Too flaky...

nityanandaz committed:
Pragmatic: filter invalid properties

nityanandaz committed:
Create testGreeterComponent

nityanandaz committed:
Create ProtobufferBuilderTests.buildService

nityanandaz committed:
Handle top-level scalar type

nityanandaz committed:
rm import

nityanandaz committed:
Resolve warning

nityanandaz committed:
Inspect Component Parameter properties

nityanandaz committed:
Update testScalarType

nityanandaz committed:
Update testGreeterComponent

nityanandaz committed:
**It's working!**

nityanandaz committed:
Update ProtobufferSemanticModelBuilder

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/inspect-component-properties

nityanandaz committed:
Create ProtobufferBuilderInternalTests

nityanandaz committed:
Delete old ProtobufferBuilder.addService

nityanandaz committed:
Merge branch 'protobuffer-generation/inspect-component-properties' into create-GRPCProtobufBuilder

nityanandaz committed:
SwiftLint

nityanandaz committed:
Catch error with "Runtime.Kind.opaque"

nityanandaz committed:
Create testHelloWorldComponent

Default inputType becomes Void

Ref.: https://github.com/Apodini/Apodini/pull/33#discussion_r541013183

nityanandaz committed:
SwiftLint

nityanandaz committed:
Merge branch 'develop' into create-GRPCProtobufBuilder

nityanandaz committed:
Add @_implementationOnly

nityanandaz committed:
Support Optionals

Ref.: https://github.com/Apodini/Apodini/pull/33#discussion_r541018208

nityanandaz committed:
Documentation

nityanandaz committed:
Change Node semantics: class

nityanandaz committed:
Adapt ProtobufferBuilderInternalTests

nityanandaz committed:
Revert "Change Node semantics: class"

This reverts commit e0b6f3592c46aca74b31a0bc59f7c57598cabd1f.

nityanandaz committed:
Fix Protocompiler error: required fields are not allowed in proto3

nityanandaz committed:
Merge branch 'develop' into create-GRPCProtobufBuilder

nityanandaz committed:
Merge branch 'create-GRPCProtobufBuilder' into protobuffer-generation/handle-recursive-type-declarations

nityanandaz committed:
Implement Tree.contains

nityanandaz committed:
Create DidFindRecursionError

TODO: Inject trap state into tree

nityanandaz committed:
Tree extensions become Node extensions

nityanandaz committed:
More nodes, less trees

nityanandaz committed:
Create tests GenericType

- Change Node.reduce semantics
- Create Node.forEach

nityanandaz committed:
Simplify TypeInfo.compatibleGenericName

nityanandaz committed:
Merge branch 'private/refactor-tree-extensions' into protobuffer-generation/handle-recursive-type-declarations

Remove Optionals where Optionals are not needed

nityanandaz committed:
Fix testRecursionSecondOrder

nityanandaz committed:
Create ProtobufferBuilderDidEncounterCircle

nityanandaz committed:
Tidy up

nityanandaz committed:
Merge branch 'protobuffer-generation/handle-recursive-type-declarations' into create-GRPCProtobufBuilder

nityanandaz committed:
Merge branch 'develop' into create-GRPCProtobufBuilder

nityanandaz committed:
Build serviceName with StringPathBuilder

nityanandaz committed:
SwiftLint

nityanandaz committed:
Always add Service at end

nityanandaz committed:
Conform ProtobufferSemanticModelBuilder to InterfaceExporter

nityanandaz committed:
SwiftLint

nityanandaz committed:
Register ProtobufferSemanticModelBuilder as interfaceExporter

nityanandaz committed:
Test with XCTVapor!

- Add WebService.prepare
- Leak Vapor.Application internally!

nityanandaz committed:
SwiftLint

lschlesinger commented on Sources/ProtobufferBuilder/EnrichedInfo.swift:
```diff
@@ -0,0 +1,56 @@
+    let propertyInfo: PropertyInfo?
+    let propertiesOffset: Int?
+    
+    var cardinality: Cardinality = .exactlyOne
```
what about optional array types, imo we should change `cardinality` to `[Cardinality]` to allow this.

lschlesinger reviewed (approved):
Looks good. As discussed, I will move your `Tree/Node`, `EnrichedInfo` and `ParticularType` as well as the `Handle...` code into the `SharedSemanticModel` to make use of it in #35

nityanandaz committed:
Merge branch 'develop' into create-GRPCProtobufBuilder

# Conflicts:
#	Sources/Apodini/WebService.swift

nityanandaz committed:
Supplement

nityanandaz committed:
Collect associated targets

moritzschuell reviewed (approved):
Great work, looks good to me 🥳 Thank you.
